src/data/fintype/basic.lean

```diff
@@ -1591,6 +1591,19 @@ not_nonempty_iff.trans is_empty_fintype
 noncomputable def fintype_of_not_infinite {α : Type*} (h : ¬ infinite α) : fintype α :=
 ((not_iff_comm.mp not_nonempty_fintype).mp h).some
 
+section
+open_locale classical
+
+/--
+Any type is (classically) either a `fintype`, or `infinite`.
+
+One can obtain the relevant typeclasses via `cases fintype_or_infinite α; resetI`.
+-/
+noncomputable def fintype_or_infinite (α : Type*) : psum (fintype α) (infinite α) :=
+if h : infinite α then psum.inr h else psum.inl (fintype_of_not_infinite h)
```

Comment:
If you make this `if i : fintype \a then psum.inl i else _` I guess the `cases fintype_or_infinite` might even preserve some nice defeqness in the case `\a` is a fintype. Is that worth doing?

Reply:
@jcommelin That doesn't seem to work -- `fintype` is a Type and not a Prop so it can't be the condition to an `if` (and even if it could, there's no reason `h` would be "the" `fintype` instance) but that seemed like a nice idea.

Reply:
Yeah, you are right.